### PR TITLE
Use doctrine/static-website-generator

### PR DESCRIPTION
**Pull request by jwage**
Use https://github.com/doctrine/static-website-generator for the basic static site generator functionality.

### CODE REVIEW

File: lib/Routing/Router.php [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        $this->routes = new RouteCollection();
+
+        $url = parse_url($site->getUrl());
+
+        $this->context = new RequestContext(
+            '',
+            'GET',
+            $url['host'],
+            $url['scheme'],
+            $url['port'],
```
Comment (SenseException) [inline lib/Routing/Router.php]: Can you make the port optional? This new version of the website doesn't allow `doctrine.website.url` without a port.

Notice: Undefined index: port in lib/Routing/Router.php on line 40

File: lib/Builder/SourceFileParametersFactory.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    public function createSourceFileParameters(
+        string $buildDir,
+        string $writePath,
+        string $string
+    ) : SourceFileParameters {
+        $parameters = [];
+
+        if (preg_match('/^\s*(?:---[\s]*[\r\n]+)(.*?)(?:---[\s]*[\r\n]+)(.*?)$/s', $string, $matches) > 0) {
```
Comment (SenseException) [inline lib/Builder/SourceFileParametersFactory.php]: This if-block should be in a private method and get a good name.

File: lib/Request/RequestCollectionProvider.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        foreach ($providers as $provider) {
+            $this->providers[get_class($provider)] = $provider;
+        }
+    }
+
+    public function getRequestCollection(string $className, string $method) : RequestCollection
+    {
+        /** @var callable $callable */
+        $callable = [$this->providers[$className], $method];
```
Comment (SenseException) [inline lib/Request/RequestCollectionProvider.php]: There's a missing handling for providers without a fitting class name.

File: lib/Routing/Router.php [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    private function createRequestContext(Site $site) : RequestContext
+    {
+        $url = parse_url($site->getUrl());
+
+        return new RequestContext(
+            '',
+            'GET',
+            $url['host'],
+            $url['scheme'],
+            $url['port'],
```
Comment (SenseException) [inline lib/Routing/Router.php]: Can you make the port optional? This new version of the website doesn't allow `doctrine.website.url` without a port.

File: templates/contributors-list.html.twig [hunk 5]
```diff
@@ -4,7 +4,9 @@
             <img src="{{ contributor.avatarUrl }}" class="rounded float-left mr-5" style="width: 50px;" />
 
             <strong>
-                <a href="https://github.com/{{ contributor.github }}/" target="_blank">{{ contributor.github }}</a>
+                <a href="{{ path('team_contributor', {github:contributor.github}) }}">
```
Comment (SenseException) [inline templates/contributors-list.html.twig]: No new window anymore?

File: templates/maintainers-list.html.twig [hunk 6]
```diff
@@ -4,8 +4,8 @@
             <img src="{{ contributor.avatarUrl }}" class="rounded float-left mr-5" style="width: 50px;" />
 
             <strong>
-                <a href="https://github.com/{{ contributor.github }}/" target="_blank">
-                    {{ contributor.teamMember.name ?? member.github }}
+                <a href="{{ path('team_contributor', {github:contributor.github}) }}">
```
Comment (SenseException) [inline templates/maintainers-list.html.twig]: No new window here too?

File: composer.json [hunk 7]
```diff
@@ -17,6 +17,7 @@
         "doctrine/inflector": "^1.3",
         "doctrine/rst-parser": "dev-master",
         "doctrine/skeleton-mapper": "dev-master",
+        "doctrine/static-website-generator": "dev-master",
```
[the remark this answers is not in the document]
Comment (jwage) [reply composer.json]: I think so. Getting the standalone pieces integrated here was one of the last steps. I still want to upgrade doctrine/doctrine-website to be a Symfony Flex app which uses https://github.com/doctrine/DoctrineSkeletonMapperBundle and https://github.com/doctrine/DoctrineStaticWebsiteGeneratorBundle

File: lib/DataSources/BlogPosts.php [hunk 8]
```diff
@@ -4,34 +4,33 @@
...
     /**
      * @return mixed[][]
      */
     public function getSourceRows() : array
     {
-        $sourceFiles = $this->sourceFileRepository->getFiles('', 'source/blog');
-
-        $reversedSourceFiles = array_reverse($sourceFiles);
+        $sourceFiles = $this->sourceFileFilesystemReader
+            ->getSourceFiles()->in('/blog/');
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/DataSources/BlogPosts.php]: `SourceFileFilesystemReader` is a part of the static website generator library now. The `in()` method is for end users to be able to filter the source files and do other things with them.

File: templates/contributors-list.html.twig (hunk 5 above)
Comment (jwage) [reply templates/contributors-list.html.twig]: The link just changed from a github link to a team member page.

File: lib/Commands/BuildDocsCommand.php [hunk 9]
```diff
@@ -31,13 +34,15 @@ protected function configure() : void
...
+                'The project to build the docs for.',
+                ''
             )
             ->addOption(
                 'v',
                 null,
                 InputOption::VALUE_REQUIRED,
-                'The project version to build the docs for.'
+                'The project version to build the docs for.',
+                ''
```
[the remark this answers is not in the document]
Comment (jwage) [reply lib/Commands/BuildDocsCommand.php]: The `env` option was orphaned. I removed it.

[the remark this answers is not in the document]
Comment (SenseException) [reply lib/Commands/BuildDocsCommand.php]: There are still 2 commands that need `env`, even though they get it injected later. But that environment variable still comes from the usage of a command option. What about if we move the option where all commands can access it? In this case `Application`:

``` php
    public function run(InputInterface $input) : int
    {
        $inputOption = new InputOption(
            'env',
            'e',
            InputOption::VALUE_REQUIRED,
            'The environment.',
            ''
        );
        $this->application->getDefinition()->addOption($inputOption);

        $this->application->add($this->buildDocsCommand);
        $this->application->add($this->buildWebsiteCommand);
        $this->application->add($this->clearBuildCacheCommand);
        $this->application->add($this->deployCommand);

        return $this->application->run($input);
    }
```

[the remark this answers is not in the document]
Comment (jwage) [reply lib/Commands/BuildDocsCommand.php]: :+1: I made this change. @SenseException can you have another look?

**Pull request merged**